File: src/service/implementations/NotificationService.cpp
#include "NotificationService.h"
#include "../repository/postgres/NotificationRepo.h"
#include "../repository/postgres/PostgresConnection.h"
#include <iostream>
#include <sstream>

NotificationService::NotificationService() 
    : notification_repo(std::make_unique<NotificationRepo>()) {
}

bool NotificationService::subscribeUser(int user_id, int product_id) {
    try {
        // Verify user exists
        auto& conn = PostgresConnection::getConnection();
        pqxx::work txn(conn);
        
        pqxx::result user_check = txn.exec_params(
            "SELECT id FROM users WHERE id = $1",
            user_id
        );
        
        if (user_check.empty()) {
            std::cerr << "User " << user_id << " not found" << std::endl;
            return false;
        }
        
        pqxx::result product_check = txn.exec_params(
            "SELECT id FROM products WHERE id = $1",
            product_id
        );
        
        if (product_check.empty()) {
            std::cerr << " Product " << product_id << " not found" << std::endl;
            return false;
        }
        
        txn.commit();
        
        // Subscribe to notifications
        return notification_repo->subscribeToNotification(user_id, product_id, "restocked");
    } catch (const std::exception& e) {
        std::cerr << "Error subscribing user: " << e.what() << std::endl;
        return false;
    }
}

bool NotificationService::unsubscribeUser(int notification_id) {
    return notification_repo->unsubscribeFromNotification(notification_id);
}

std::vector<domain::Notification> NotificationService::getUserNotifications(int user_id) {
    return notification_repo->getUserNotifications(user_id);
}

std::vector<domain::Notification> NotificationService::getProductSubscribers(int product_id) {
    return notification_repo->getProductSubscribers(product_id);
}

bool NotificationService::sendRestockNotifications(int product_id, int stock) {
    try {
        if (stock <= 0) {
            std::cout << "⚠️  Product " << product_id << " is out of stock, not sending notifications" << std::endl;
            return true;
        }
        
        // Get all subscribers for this product who haven't been notified yet
        std::vector<domain::Notification> subscribers = notification_repo->getProductSubscribers(product_id);
        
        if (subscribers.empty()) {
            std::cout << "ℹ️  No subscribers for product " << product_id << std::endl;
            return true;
        }
        
        std::cout << "📢 Sending restock notifications to " << subscribers.size() << " subscribers of product " << product_id << std::endl;
        
        auto& conn = PostgresConnection::getConnection();
        
        for (const auto& sub : subscribers) {
            if (sub.is_sent) {
                continue; // Skip already notified subscribers
            }
            
            std::string message = "Product back in stock! Check it out now.";
            
            if (!sendNotification(sub.id, message)) {
                std::cerr << "Failed to send notification " << sub.id << " to user " << sub.user_id << std::endl;
            }
        }
        
        return true;
    } catch (const std::exception& e) {
        std::cerr << "Error sending restock notifications: " << e.what() << std::endl;
        return false;
    }
}

bool NotificationService::sendNotification(int notification_id, const std::string& message) {
    try {
        // Get notification details
        domain::Notification notif = notification_repo->getNotificationById(notification_id);
        if (notif.id == 0) {
            std::cerr << "Notification " << notification_id << " not found" << std::endl;
            return false;
        }
        
        // Get user preferences
        domain::NotificationPreference prefs = notification_repo->getNotificationPreference(notif.user_id);
        
        // Create notification log entry
        domain::NotificationLog log;
        log.notification_id = notification_id;
        log.user_id = notif.user_id;
        log.product_id = notif.product_id;
        log.notification_type = notif.type_str;
        log.message = message;
        log.status = "pending";
        log.retry_count = 0;
        log.max_retries = 3;
        
        int log_id = notification_repo->createNotificationLog(log);
        if (log_id < 0) {
            std::cerr << "Failed to create notification log" << std::endl;
            return false;
        }
        
        // Simulate notification sending (in real app, send email, push, etc.)
        bool send_success = simulateSendNotification(notif.user_id, notif.product_id, message, prefs);
        
        if (send_success) {
            // Mark notification as sent
            notification_repo->markNotificationAsSent(notification_id);
            notification_repo->updateNotificationLogStatus(log_id, "sent");
            
            std::cout << "Notification sent to user " << notif.user_id 
                     << " for product " << notif.product_id << std::endl;
            return true;
        } else {
            // Mark as failed in log
            notification_repo->updateNotificationLogStatus(log_id, "failed", "Failed to send notification");
            
            std::cerr << " Failed to send notification to user " << notif.user_id << std::endl;
            return false;
        }
    } catch (const std::exception& e) {
        std::cerr << "Error in sendNotification: " << e.what() << std::endl;
        return false;
    }
}

bool NotificationService::simulateSendNotification(int user_id, int product_id, 
                                                   const std::string& message,
                                                   const domain::NotificationPreference& prefs) {
    try {
        std::ostringstream ss;
        ss << "📧 [NOTIFICATION LOG] User " << user_id 
           << " - Product " << product_id 
           << " - Message: " << message << std::endl;
        ss << "   ├─ Email: " << (prefs.email_enabled ? "enabled" : "disabled") << std::endl;
        ss << "   ├─ Push: " << (prefs.push_enabled ? "enabled" : "disabled") << std::endl;
        ss << "   ├─ SMS: " << (prefs.sms_enabled ? "enabled" : "disabled") << std::endl;
        ss << "   └─ In-App: " << (prefs.in_app_enabled ? "enabled" : "disabled") << std::endl;
        
        std::cout << ss.str();
        
        // In a real implementation:
        // - Send email via SMTP if email_enabled
        // - Send push via Firebase/APNs if push_enabled
        // - Send SMS via Twilio if sms_enabled
        // - Create in-app notification if in_app_enabled
        
        return true; // Simulate successful send
    } catch (const std::exception& e) {
        std::cerr << "Error simulating notification: " << e.what() << std::endl;
        return false;
    }
}

std::vector<domain::NotificationLog> NotificationService::getNotificationLogs(int user_id) {
    return notification_repo->getNotificationLogs(user_id);
}

std::vector<domain::NotificationLog> NotificationService::getFailedNotifications() {
    return notification_repo->getFailedNotifications();
}

bool NotificationService::retryFailedNotification(int log_id) {
    try {
        std::vector<domain::NotificationLog> logs = notification_repo->getFailedNotifications();
        
        domain::NotificationLog target_log;
        for (const auto& log : logs) {
            if (log.id == log_id) {
                target_log = log;
                break;
            }
        }
        
        if (target_log.id == 0) {
            std::cerr << "Failed notification " << log_id << " not found or already recovered" << std::endl;
            return false;
        }
        
        if (target_log.retry_count >= target_log.max_retries) {
            std::cerr << "Max retries reached for notification " << log_id << std::endl;
            return false;
        }
        
        // Increment retry count
        notification_repo->incrementRetryCount(log_id);
        
        // Try sending again
        std::string message = target_log.message;
        if (sendNotification(target_log.notification_id, message)) {
            notification_repo->updateNotificationLogStatus(log_id, "retried");
            std::cout << "Notification " << log_id << " retried successfully" << std::endl;
            return true;
        } else {
            std::cerr << "Retry failed for notification " << log_id << std::endl;
            return false;
        }
    } catch (const std::exception& e) {
        std::cerr << "Error retrying notification: " << e.what() << std::endl;
        return false;
    }
}

bool NotificationService::updateUserPreferences(int user_id, bool email, bool push, bool sms, bool in_app) {
    try {
        domain::NotificationPreference pref = notification_repo->getNotificationPreference(user_id);
        pref.email_enabled = email;
        pref.push_enabled = push;
        pref.sms_enabled = sms;
        pref.in_app_enabled = in_app;
        
        return notification_repo->updateNotificationPreference(pref);
    } catch (const std::exception& e) {
        std::cerr << " Error updating user preferences: " << e.what() << std::endl;
        return false;
    }
}

domain::NotificationPreference NotificationService::getUserPreferences(int user_id) {
    return notification_repo->getNotificationPreference(user_id);
}

File: frontend/src/services/notificationService.ts
const API_BASE_URL = process.env.REACT_APP_API_BASE_URL || 'http://localhost:8080';
const API_URL = `${API_BASE_URL}/api`;

export interface ProductNotification {
  id: number;
  product_id: number;
  user_id: number;
  type: string;
  is_sent: boolean;
  created_at: string;
  sent_at?: string;
}

export interface NotificationLog {
  id: number;
  notification_id: number;
  user_id: number;
  product_id: number;
  type: string;
  message: string;
  status: 'pending' | 'sent' | 'failed' | 'retried';
  retry_count: number;
  max_retries: number;
  error_message?: string;
  sent_at?: string;
  created_at: string;
}

export interface NotificationPreference {
  id: number;
  user_id: number;
  email_enabled: boolean;
  push_enabled: boolean;
  sms_enabled: boolean;
  in_app_enabled: boolean;
  created_at: string;
  updated_at: string;
}

class NotificationService {
  /**
   * Subscribe a user to product restock notifications
   */
  async subscribeToProduct(userId: number, productId: number): Promise<ProductNotification> {
    try {
      const response = await fetch(`${API_URL}/notifications/subscribe`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ user_id: userId, product_id: productId })
      });
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${response.statusText}`);
      }
      const data = await response.json();
      console.log('✅ Subscribed to product notifications:', data);
      return data;
    } catch (error) {
      console.error('❌ Error subscribing to product:', error);
      throw error;
    }
  }

  /**
   * Unsubscribe from a specific notification
   */
  async unsubscribeFromNotification(notificationId: number): Promise<{ status: string; message: string }> {
    try {
      const response = await fetch(`${API_URL}/notifications/${notificationId}`, {
        method: 'DELETE'
      });
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${response.statusText}`);
      }
      const data = await response.json();
      console.log('✅ Unsubscribed from notifications:', data);
      return data;
    } catch (error) {
      console.error('❌ Error unsubscribing:', error);
      throw error;
    }
  }

  /**
   * Get all notifications for a user
   */
  async getUserNotifications(userId: number): Promise<ProductNotification[]> {
    try {
      const response = await fetch(`${API_URL}/notifications/user/${userId}`);
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${response.statusText}`);
      }
      const data = await response.json();
      console.log('✅ Retrieved user notifications:', data);
      return data;
    } catch (error) {
      console.error('❌ Error getting user notifications:', error);
      throw error;
    }
  }

  /**
   * Get all subscribers for a product
   */
  async getProductSubscribers(productId: number): Promise<ProductNotification[]> {
    try {
      const response = await fetch(`${API_URL}/notifications/product/${productId}`);
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${response.statusText}`);
      }
      const data = await response.json();
      console.log('✅ Retrieved product subscribers:', data);
      return data;
    } catch (error) {
      console.error('❌ Error getting product subscribers:', error);
      throw error;
    }
  }

  /**
   * Get notification logs for a user
   */
  async getNotificationLogs(userId: number): Promise<NotificationLog[]> {
    try {
      const response = await fetch(`${API_URL}/notifications/logs/user/${userId}`);
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${response.statusText}`);
      }
      const data = await response.json();
      console.log('✅ Retrieved notification logs:', data);
      return data;
    } catch (error) {
      console.error('❌ Error getting notification logs:', error);
      throw error;
    }
  }

  /**
   * Get all failed notifications in the system
   */
  async getFailedNotifications(): Promise<NotificationLog[]> {
    try {
      const response = await fetch(`${API_URL}/notifications/logs/status/failed`);
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${response.statusText}`);
      }
      const data = await response.json();
      console.log('✅ Retrieved failed notifications:', data);
      return data;
    } catch (error) {
      console.error('❌ Error getting failed notifications:', error);
      throw error;
    }
  }

  /**
   * Retry sending a failed notification
   */
  async retryFailedNotification(logId: number): Promise<{ status: string; message: string }> {
    try {
      const response = await fetch(`${API_URL}/notifications/logs/${logId}/retry`, {
        method: 'POST'
      });
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${response.statusText}`);
      }
      const data = await response.json();
      console.log('✅ Notification retry initiated:', data);
      return data;
    } catch (error) {
      console.error('❌ Error retrying notification:', error);
      throw error;
    }
  }

  /**
   * Update user notification preferences
   */
  async updateNotificationPreferences(
    userId: number,
    preferences: Partial<Omit<NotificationPreference, 'id' | 'user_id' | 'created_at' | 'updated_at'>>
  ): Promise<{ status: string; message: string }> {
    try {
      const response = await fetch(`${API_URL}/notifications/preferences/${userId}`, {
        method: 'PUT',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(preferences)
      });
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${response.statusText}`);
      }
      const data = await response.json();
      console.log('✅ Notification preferences updated:', data);
      return data;
    } catch (error) {
      console.error('❌ Error updating preferences:', error);
      throw error;
    }
  }

  /**
   * Get user notification preferences
   */
  async getNotificationPreferences(userId: number): Promise<NotificationPreference> {
    try {
      const response = await fetch(`${API_URL}/notifications/preferences/${userId}`);
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${response.statusText}`);
      }
      const data = await response.json();
      console.log('✅ Retrieved notification preferences:', data);
      return data;
    } catch (error) {
      console.error('❌ Error getting preferences:', error);
      throw error;
    }
  }

  /**
   * Check if user is subscribed to a product
   */
  async isUserSubscribed(userId: number, productId: number): Promise<boolean> {
    try {
      const notifications = await this.getUserNotifications(userId);
      return notifications.some(n => n.product_id === productId && !n.is_sent);
    } catch {
      return false;
    }
  }

  /**
   * Get subscription for a specific product
   */
  async getProductSubscription(userId: number, productId: number): Promise<ProductNotification | null> {
    try {
      const notifications = await this.getUserNotifications(userId);
      return notifications.find(n => n.product_id === productId) || null;
    } catch {
      return null;
    }
  }

  /**
   * Get unread notification count for user
   */
  async getUnreadNotificationCount(userId: number): Promise<number> {
    try {
      const logs = await this.getNotificationLogs(userId);
      return logs.filter(l => l.status === 'sent' && !l.sent_at).length;
    } catch {
      return 0;
    }
  }
}

export default new NotificationService();
